qlib/data/storage/storage.py

```diff
@@ -0,0 +1,420 @@
+        )
+
+    def check_exists(self) -> bool:
+        '''check if storage(uri) exists, if not exists: return False'''
+        raise NotImplementedError("Subclass of BaseStorage must implement `check_exists` method")
+
+"""
+
+
+class StorageMeta(type):
+    """unified management of raise when storage is not exists"""
+
+    def __new__(cls, name, bases, dict):
+        class_obj = type.__new__(cls, name, bases, dict)
+
+        # The calls to __iter__ and __getitem__ do not pass through __getattribute__.
+        # In order to throw an exception before calling __getitem__, use the metaclass
+        _getitem_func = getattr(class_obj, "__getitem__")
+
+        def _getitem(obj, item):
+            getattr(obj, "_check")()
+            try:
+                res = _getitem_func(obj, item)
+            except Exception as e:
+                raise ValueError(f"{obj.raise_info}\n\tStorage exception info: {str(e)}")
+            return res
+
+        setattr(class_obj, "__getitem__", _getitem)
+        return class_obj
+
+
+class BaseStorage(metaclass=StorageMeta):
+    @property
+    def storage_name(self) -> str:
+        return re.findall("[A-Z][^A-Z]*", self.__class__.__name__)[-2].lower()
+
+    @property
+    def raise_info(self):
+        parameters_info = [
+            f"{_k}={_v}"
+            for _k, _v in self.__dict__.items()
+            if not isinstance(_v, (dict,)) or (hasattr(_v, "__len__") and len(_v) < 3)
+        ]
+        return f"{self.storage_name.lower()} not exists, storage parameters: {parameters_info}"
+
+    def check_exists(self) -> bool:
+        """check if storage(uri) exists, if not exists: return False"""
+        raise NotImplementedError("Subclass of BaseStorage must implement `check_exists` method")
+
+    def clear(self) -> None:
+        """clear storage"""
+        raise NotImplementedError("Subclass of BaseStorage must implement `clear` method")
+
+    def __len__(self) -> 0:
+        return len(self.data) if self.check_exists() else 0
+
+    def __getitem__(self, item: Union[slice, Union[int, InstKT]]):
+        raise NotImplementedError(
+            "Subclass of BaseStorage must implement `__getitem__(i: Union[int, InstKT])`/`__getitem__(s: slice)`  method"
+        )
+
+    def _check(self):
+        if not self.check_exists():
+            raise ValueError(self.raise_info)
+
+    def __getattribute__(self, item):
+        if item == "data":
+            self._check()
+        try:
+            res = super(BaseStorage, self).__getattribute__(item)
+        except Exception as e:
+            raise ValueError(f"{self.raise_info}\n\tStorage exception info: {str(e)}")
+        return res
+
+
+class CalendarStorage(BaseStorage):
+    """
+    The behavior of CalendarStorage's methods and List's methods of the same name remain consistent
+    """
+
+    def __init__(self, freq: str, future: bool, uri: str, **kwargs):
+        self.freq = freq
+        self.future = future
+        self.uri = uri
```

Comment:
The assignment of the URI property looks a little misleading (user will think `self.uri` should represent the `provider_uri`).

Reply:
remove this attribute